operator/internal/lifecycle/client.go

```diff
@@ -0,0 +1,387 @@
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+package lifecycle
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	k8sapierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// Cluster is a generic interface for a pointer to a Kubernetes object
+// that represents a cluster.
+type Cluster[T any] interface {
+	client.Object
+	*T
+}
+
+// NewClusterObject creates a new instance of a typed cluster object.
+func NewClusterObject[T any, U Cluster[T]]() U {
+	var t T
+	return U(&t)
+}
+
+// NewResourceClient creates a new instance of a ResourceClient for managing resources.
+func NewResourceClient[T any, U Cluster[T]](mgr ctrl.Manager, resourcesFn ResourceManagerFactory[T, U]) *ResourceClient[T, U] {
+	ownershipResolver, statusUpdater, nodePoolRenderer, simpleResourceRenderer := resourcesFn(mgr)
+	return &ResourceClient[T, U]{
+		client:                 mgr.GetClient(),
+		scheme:                 mgr.GetScheme(),
+		mapper:                 mgr.GetRESTMapper(),
+		ownershipResolver:      ownershipResolver,
+		statusUpdater:          statusUpdater,
+		nodePoolRenderer:       nodePoolRenderer,
+		simpleResourceRenderer: simpleResourceRenderer,
+	}
+}
+
+// ResourceClient is a client used to manage dependent resources,
+// both simple and node pools, for a given cluster type.
+type ResourceClient[T any, U Cluster[T]] struct {
+	client                 client.Client
+	scheme                 *runtime.Scheme
+	mapper                 meta.RESTMapper
+	ownershipResolver      OwnershipResolver[T, U]
+	statusUpdater          ClusterStatusUpdater[T, U]
+	nodePoolRenderer       NodePoolRenderer[T, U]
+	simpleResourceRenderer SimpleResourceRenderer[T, U]
+}
+
+// PatchNodePoolSet updates a StatefulSet for a specific node pool.
+func (r *ResourceClient[T, U]) PatchNodePoolSet(ctx context.Context, owner U, set *appsv1.StatefulSet) error {
+	return r.patchOwnedResource(ctx, owner, set)
+}
+
+// SetClusterStatus sets the status of the given cluster.
+func (r *ResourceClient[T, U]) SetClusterStatus(cluster U, status ClusterStatus) bool {
+	return r.statusUpdater.Update(cluster, status)
+}
+
+type gvkObject struct {
+	gvk schema.GroupVersionKind
+	nn  types.NamespacedName
+}
+
+// SyncAll synchronizes the simple resources associated with the given cluster,
+// cleaning up any resources that should no longer exist.
```

Comment:
Can you include the phrase "garbage collect" or "GC" in here or even in a comment in the struct itself. We and Kubernetes refer to the process as garbage collection, adding the key words here will make it a lot easier to find this for those that aren't intimately familiar with this section of the code base.